flintrock.py

```diff
@@ -533,31 +528,80 @@ def launch_ec2(
         vpc_id=vpc_id,
         region=region)
 
+    num_instances = num_slaves + 1
     try:
         connection = boto.ec2.connect_to_region(region_name=region)
 
-        print("Launching {c} instances...".format(c=num_slaves + 1))
+        if spot_price is not None:
+            print("Requesting {c} spot instances at a max price of {p}".format(c=num_instances, p=spot_price))
+
+            instance_reqs = connection.request_spot_instances(
+                price=spot_price,
+                image_id=ami,
+                count=(num_instances),
+                key_name=key_name,
+                instance_type=instance_type,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            pending_request_ids = [req.id for req in instance_reqs]
+
+            try:
+
+                while len(pending_request_ids) > 0:
+                    time.sleep(30)
 
-        reservation = connection.run_instances(
-            image_id=ami,
-            min_count=(num_slaves + 1),
-            max_count=(num_slaves + 1),
-            key_name=key_name,
-            instance_type=instance_type,
-            placement=availability_zone,
-            security_group_ids=[sg.id for sg in security_groups],
-            subnet_id=subnet_id,
-            placement_group=placement_group,
-            tenancy=tenancy,
-            ebs_optimized=ebs_optimized,
-            instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
+                    spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
+                    pending_request_ids = [req.id for req in spot_requests if req.state != 'active']
 
-        time.sleep(10)  # AWS metadata eventual consistency tax.
+                    if len(pending_request_ids) > 0:
+                        print("{grant} of {req} instances granted, waiting longer".format(
+                            grant=num_instances - len(pending_request_ids), req=num_instances))
+
+                print("All {num_instances} granted".format(num_instances=num_instances))
+                spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
```

Comment:
I'm sorry, I'm not sure how that made it through; fixing now.

On Tue, Oct 27, 2015 at 2:44 PM, Nicholas Chammas <notifications@github.com>
wrote:

> In flintrock.py
> <https://github.com/nchammas/flintrock/pull/35#discussion_r43167840>:
>
> >
> > -        time.sleep(10)  # AWS metadata eventual consistency tax.
> > +                    if len(pending_request_ids) > 0:
> > +                        print("{grant} of {req} instances granted, waiting longer".format(
> > +                            grant=num_instances - len(pending_request_ids), req=num_instances))
> > +
> > +                print("All {num_instances} granted".format(num_instances=num_instances))
> > +                spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
>
> If I'm not mistaken, pending_request_ids will be empty here because they
> are all fulfilled, so this won't work. You probably want to reuse
> instance_reqs from earlier, no?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nchammas/flintrock/pull/35/files#r43167840>.
>